### PR TITLE
Feature/standard options pattern 2

### PR DESCRIPTION
**Pull request by sgryphon**
Cleans up some of the Options pattern in the Extensions Logging to more follow the way things like OptionsMonitor<>, etc work.

Options pattern is meant for configuration settings, mostly passed in from external sources, e.g. appsettigns.json, enviornment variables, command line, etc, so is limited to values only.

Configuration for things like callbacks, which can be done in code, are in a separate function.

It includes an example that add the logger provider with a single line of code (plus a using statement), to show simple usage; and keeps the more advanced using (with callbacks, etc) as a separate example.

### CODE REVIEW

**Comment by sgryphon** [comment]
I have kept the "ShipTo" as a separate child object of the main configuration, although this just means more lines in the JSON configuration, and a longer name to override from the command line.

It may be easier for the Options class to simply declare all properties at the top level, and then have intelligent defaults, e.g. if CloudId is defined, then configure as for cloud even if ConnectionPoolType is not defined (assume Unknown is actually Cloud).

Another option is to treat prefixes as the hierarchy, e.g.

```
 "Logging":  {
    "Elasticsearch": {
      "Cloud": {
        "Id": "...",
        "ApiKey": "..."
      }
    }
  }
```

As command line / environment variables, these would be like `--Logging:Elasticsearch:Cloud:ApiKey "..."`

File: src/Elasticsearch.Extensions.Logging/ShipToOptions.cs [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+		/// Gets or sets the API Key, where connection pool type is Cloud, and authenticating via API Key.
+		/// </summary>
+		public string ApiKey { get; set; } = "";
+
+		/// <summary>
+		/// Gets or sets the cloud ID, where connection pool type is Cloud.
+		/// </summary>
+		public string CloudId { get; set; } = "";
+
+   		/// <summary>
```
Comment (Mpdreamz) [inline src/Elasticsearch.Extensions.Logging/ShipToOptions.cs:22]: spaces

File: src/Elasticsearch.Extensions.Logging/ShipToOptions.cs [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+		/// <summary>
+		/// Gets or sets the password, where connection pool type is Cloud, and authenticating via username/password.
+		/// </summary>
+		public string Password { get; set; } = "";
+
+		/// <summary>
+		/// Gets or sets the username, where connection pool type is Cloud, and authenticating via username/password.
+		/// </summary>
+		public string Username { get; set; } = "";
+    }
```
Comment (Mpdreamz) [inline src/Elasticsearch.Extensions.Logging/ShipToOptions.cs:43]: spaces

File: tests/Elasticsearch.Extensions.Logging.IntegrationTests/TestChannelSetup.cs [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+// See the LICENSE file in the project root for more information
+
+using System;
+using Elastic.Ingest;
+
+namespace Elasticsearch.Extensions.Logging.IntegrationTests
+{
+    internal class TestChannelSetup : IChannelSetup
+    {
+        Action<ElasticsearchChannelOptions<LogEvent>> _configureChannel;
```
Comment (Mpdreamz) [inline tests/Elasticsearch.Extensions.Logging.IntegrationTests/TestChannelSetup.cs:12]: spaces

File: src/Elasticsearch.Extensions.Logging/IChannelSetup.cs [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+// See the LICENSE file in the project root for more information
+
+using System;
+using Elastic.Ingest;
+
+namespace Elasticsearch.Extensions.Logging
+{
+    public interface IChannelSetup
+    {
+        void ConfigureChannel(ElasticsearchChannelOptions<LogEvent> channelConfiguration);
```
Comment (Mpdreamz) [inline src/Elasticsearch.Extensions.Logging/IChannelSetup.cs:12]: spaces

File: examples/console-with-extensions-logging/Worker.cs [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+using System.Security.Principal;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Hosting;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Logging.Configuration;
+
+namespace ConsoleExample
+{
+    public class Worker : BackgroundService
```
Comment (Mpdreamz) [inline examples/console-with-extensions-logging/Worker.cs:15]: spaces

**Review by Mpdreamz — APPROVED** [review]
Thank you @sgryphon 👍 this looks great!

I would like to keep the `ShipTo` configuration isolated for now. Thanks for leaving that as is. 

LGTM, i added a couple of small nits around spaces. Out of curiosity: Are you using VS2019 or Rider? These should be picking up the `.editorconfig` settings.

File: src/Elasticsearch.Extensions.Logging/ShipToOptions.cs (hunk 1 above)
Comment (sgryphon) [reply src/Elasticsearch.Extensions.Logging/ShipToOptions.cs]: I'm usually using Rider and VS Code these days, mostly on Ubuntu, although sometimes on Windows. (Both should support editor config though).

I cleaned the files in Rider, and then also did a clean on all the other files I touched in the branch/PR.  This did reorganise some of the files, as my code cleanup is configured to put members in alphabetical order (with public before private).

The specific configuration I use for Rider/Resharper is here: https://github.com/sgryphon/csharp-layout  (I usually use the StyleCop layout, allowing the standard rules to be turned on for checking during build, etc).

**Comment by sgryphon** [comment]
Code review changes are done, and everything looks green, although I don't have access to merge. @Mpdreamz did you want to merge (the extensions logging isn't released yet; I did see the project list for other things to fix up)?

**Comment by sgryphon** [comment]
@Mpdreamz I see you merged some other branches, so this one had a conflict. I have fixed that now, so good to merge.

File: README.md [hunk 6]
```diff
@@ -63,6 +63,25 @@ Foundational project that contains a full C# representation of ECS. [Learn more.
 
 ## Logging
 
+### [Elasticsearch.Extensions.Loggging](src/Elasticsearch.Extensions.Logging/ReadMe.md)
+
+Elasticsearch-Logstash-Kibana (ELK) stack logger provider for Microsoft.Extensions.Logging.
```
Comment (Mpdreamz) [inline README.md:68]:
```suggestion
Elastic Stack (ELK) logger provider for `Microsoft.Extensions.Logging`.
```

**Review by Mpdreamz — APPROVED** [review]
LGTM, thank you for taking this on @sgryphon very much appreciated.

**Pull request merged**